src/network/http.cpp

```diff
@@ -31,198 +31,299 @@ void http_dispose() { }
+        response->size = writeBuffer.length;
+    }
+
+    curl_easy_cleanup(curl);
+
+    return response;
+}
 
-		char contentLengthHeaderValue[64];
-		snprintf(contentLengthHeaderValue, sizeof(contentLengthHeaderValue), "Content-Length: %zu", readBuffer.length);
-		headers = curl_slist_append(headers, contentLengthHeaderValue);
+void http_request_async(const http_request_t *request, void (*callback)(http_response_t*))
+{
+    struct TempThreadArgs {
+        http_request_t request;
+        void (*callback)(http_response_t*);
+    };
+
+    TempThreadArgs *args = (TempThreadArgs*)malloc(sizeof(TempThreadArgs));
+    args->request.url = _strdup(request->url);
+    args->request.method = request->method;
+
+    if (request->type == HTTP_DATA_JSON) {
+        args->request.root = json_deep_copy(request->root);
+    } else {
+        char* bodyCopy = (char*) malloc(request->size);
+        memcpy(bodyCopy, request->body, request->size);
+        args->request.body = bodyCopy;
+    }
+
+    args->request.type = request->type;
+    args->request.size = request->size;
+    args->request.tag = request->tag;
+    args->callback = callback;
+
+    SDL_Thread *thread = SDL_CreateThread([](void *ptr) -> int {
+        TempThreadArgs *args = (TempThreadArgs*)ptr;
+
+        http_response_t *response = http_request(&args->request);
+        args->callback(response);
+
+        free((char*)args->request.url);
+
+        if (args->request.type == HTTP_DATA_JSON) {
+            json_decref((json_t*) args->request.root);
+        } else {
+            free(args->request.body);
+        }
+
+        free(args);
+        return 0;
+    }, NULL, args);
+
+    if (thread == NULL) {
+        log_error("Unable to create thread!");
+        callback(NULL);
+    } else {
+        SDL_DetachThread(thread);
+    }
+}
 
-		curl_easy_setopt(curl, CURLOPT_POSTFIELDS, readBuffer.ptr);
-	}
+void http_request_dispose(http_response_t *response)
+{
+    if (response->type == HTTP_DATA_JSON && response->root != NULL)
+        json_decref(response->root);
+    else if (response->type == HTTP_DATA_RAW && response->body != NULL)
+        free(response->body);
 
-	curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, request->method);
-	curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
-	curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, true);
-	curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, true);
-	curl_easy_setopt(curl, CURLOPT_USERAGENT, OPENRCT2_USER_AGENT);
-	curl_easy_setopt(curl, CURLOPT_URL, request->url);
-	curl_easy_setopt(curl, CURLOPT_WRITEDATA, &writeBuffer);
-	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, http_request_write_func);
+    free(response);
+}
 
-	curlResult = curl_easy_perform(curl);
+const char *http_get_extension_from_url(const char *url)
+{
+    const char *extension = strrchr(url, '.');
+
+    // Assume a save file by default if no valid extension can be determined
+    if (extension == NULL || strchr(extension, '/') != NULL) {
+        return ".sv6";
```

Comment:
This function is very generic given its in http.c and is just called `http_get_extension_from_url` - so its strange that it has somewhat specific to .sv6. That fallback should go in the caller.